src/Orleans.Core.Abstractions/Placement/PlacementAttribute.cs

```diff
@@ -69,4 +70,75 @@ public sealed class ActivationCountBasedPlacementAttribute : PlacementAttribute
         {
         }
     }
+
+    /// <summary>
+    /// Marks a grain class as using the <c>SiloPlacement</c> policy.
+    /// </summary>
+    /// <remarks>
+    /// This indicates the grain should always be placed on the target silo.
+    /// </remarks>
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = false)]
+    public sealed class SiloServicePlacementAttribute : PlacementAttribute
+    {
+        public SiloServicePlacementAttribute() :
+            base(SiloServicePlacement.Singleton)
+        {
+        }
+    }
+
+    public static class SiloServicePlacementKeyFormat
+    {
+        private const char Separator = '.';
+        private static readonly char[] Separators = { Separator };
+
+        public static bool TryParsePrimaryKey(string primaryKey, out string key)
+        {
+            key = default(string);
+            int index = primaryKey.IndexOfAny(Separators);
+            if (index == -1)
+                return false;
+            if (index == primaryKey.Length - 1)
+            {
+                key = string.Empty;
+                return true;
+            }
+            key = primaryKey.Substring(index + 1);
+            return true;
+        }
+
+        internal static bool TryParsePrimaryKey(IEnumerable<SiloAddress> silos, string primaryKey, out Tuple<SiloAddress, string> parsedKey)
+        {
+            parsedKey = default(Tuple<SiloAddress, string>);
+            foreach (SiloAddress silo in silos)
+            {
+                string siloTag = BuildSiloTag(silo);
+                if (primaryKey.StartsWith(siloTag))
+                {
+                    parsedKey = Tuple.Create(silo, primaryKey.Substring(siloTag.Length));
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        internal static string BuildPrimaryKey(SiloAddress silo, string key)
+        {
+            return $"{BuildSiloTag(silo)}{key}";
+        }
+
+
+        private static string BuildSiloTag(SiloAddress silo)
+        {
+            return $"{silo.GetConsistentHashCode()}{Separator}";
```

Comment:
Why not use a parseable SiloAddress? I think the intent would be more obvious (debugging, too)

Reply:
I wanted to use something that I was confident wouldn't have the separator, else I'd need to normalize.  I'll switch to normalized parseable string.